### PR TITLE
feat(dspark): add Ascend NPU support for Qwen3.5-4B DSpark training

### PR DESCRIPTION
**Pull request by curnane-lab**
## Summary

This PR adds Ascend NPU training support for DSpark on **Qwen3.5-4B**.

> **Note on scope:** The branch currently contains two commits. The first commit (`a2f18ea`, "feat: DSpark trainer") is borrowed from the preceding DSpark trainer PR and provides the base DSpark implementation. **This PR's own incremental change is the second commit (`5776d51`), which adds the NPU example script and the `flex_attention` -> SDPA fallback in the trainer.**

### 1. NPU training launcher
- `examples/run_qwen3.5_4b_dspark_online_npu.sh`
  - Sets `ASCEND_RT_VISIBLE_DEVICES` and `PYTORCH_NPU_ALLOC_CONF`.
  - Uses `--attention-backend sdpa` and `--target-model-backend hf` (HF backend always surfaces `last_hidden_states`, which DSpark's L1 / confidence losses require).
  - Uses HCCL via `torchrun --standalone`.

### 2. Trainer NPU fallback
- `scripts/train_dspark.py`
  - Auto-detects Ascend NPU and falls back from `flex_attention` to `sdpa` when the default backend would fail on NPU.

## DSpark background (for context)

DSpark = SpecForge's DFlash block-diffusion drafter + EAGLE-style Markov & confidence heads, trained with:
- Cross-entropy against ground-truth next tokens.
- L1 distribution distillation using the target model's final hidden state.
- Confidence-head BCE against the empirical per-token accept rate.

The base trainer implementation is in the preceding commit (`a2f18ea`). This PR only layers the NPU enablement on top.

## Usage

```bash
export TARGET_MODEL_PATH=/path/to/Qwen3.5-4B
export TRAIN_DATA_PATH=/path/to/train.jsonl
bash examples/run_qwen3.5_4b_dspark_online_npu.sh 0,1,2,3,4,5,6,7
```

### Hardware

8× Ascend 910C NPU

### Target model

[Qwen3.5-4B](https://huggingface.co/Qwen/Qwen3.5-4B)

### Training data preparation

Qwen3.5-4B (target model) with the **HF backend** is used to regenerate responses for **10K samples** from [Open-PerfectBlend](https://huggingface.co/datasets/mlabonne/open-perfectblend). The regenerated QA pairs serve as training data for dspark draft models.

dspark_npu

### Loss Configuration

DSpark is trained with a combination of:

- Cross-entropy loss
- L1 distribution distillation loss
- Confidence BCE loss

The loss weights used for the accuracy run are:

| Loss | Weight |
| --- | --- |
| CE loss | `0.1` |
| L1 distillation loss | `0.9` |
| Confidence head loss | `1.0` |

### Reproduction Command

```bash
TARGET_MODEL_PATH=/path/to/Qwen3.5-4B \
TRAIN_DATA_PATH=/path/to/distilled_train_data.jsonl \
bash scripts/train_dspark_npu.sh 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
```

The script launches:

```bash
torchrun \
  --standalone \
  --nproc_per_node <num_devices> \
  scripts/train_dspark.py \
  --target-model-path "$TARGET_MODEL_PATH" \
  --draft-config-path configs/qwen3.5-4b-dspark.json \
  --train-data-path "$TRAIN_DATA_PATH" \
  --output-dir outputs/qwen3.5-4b-dspark-npu \
  --num-epochs 10 \
  --batch-size 1 \
  --accumulation-steps 4 \
  --learning-rate 6e-4 \
  --warmup-ratio 0.04 \
  --max-grad-norm 1.0 \
  --max-length 1024 \
  --chat-template qwen3.5 \
  --attention-backend sdpa \
  --num-anchors 186 \
  --loss-decay-gamma 7.0 \
  --target-model-backend hf \
  --block-size 16 \
  --embedding-key model.language_model.embed_tokens.weight \
  --markov-rank 256 \
  --enable-confidence-head \
  --confidence-head-with-markov \
  --ce-loss-alpha 0.1 \
  --l1-loss-alpha 0.9 \
  --confidence-head-alpha 1.0 \
  --trust-remote-code
```

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [x] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by jiapingW** [comment]
Hi, now we have two branch. The `main` branch supports training with separate training and push, while the `colocated` branch retains both the previous colocated and training-pull separation training. This PR should be submitted to the `colocated` branch. Thanks. @curnane-lab

**Comment by curnane-lab** [comment]
> Hi, now we have two branch. The `main` branch supports training with separate training and push, while the `colocated` branch retains both the previous colocated and training-pull separation training. This PR should be submitted to the `colocated` branch. Thanks. @curnane-lab

@jiapingW Got it — will retarget to `colocated`. One dependency to sort out first:

This PR is stacked on #613: we share its 3 commits (`9facb330`, `306e1966`, `b2ba9aa5`), and my increment is only the last 2 (NPU launcher + `flex_attention`→`sdpa` fallback). #613 targets `dataflow-up-16-sgl0.5.14` and hasn't been asked to move, so if I retarget #617 alone the diff still carries #613's commits and the stack breaks.

I checked `colocated`: it already has the same `target_engine` capture-policy layout and the sglang 0.5.14 changes from #652, so #613 should rebase onto it cleanly.

Maybe #613 retargets to `colocated` first (cc @maocheng23), then I rebase #617 and follow.

**Comment by jiapingW** [comment]
> > Hi, now we have two branch. The `main` branch supports training with separate training and push, while the `colocated` branch retains both the previous colocated and training-pull separation training. This PR should be submitted to the `colocated` branch. Thanks. @curnane-lab
> 
> @jiapingW Got it — will retarget to `colocated`. One dependency to sort out first:
> 
> This PR is stacked on #613: we share its 3 commits (`9facb330`, `306e1966`, `b2ba9aa5`), and my increment is only the last 2 (NPU launcher + `flex_attention`→`sdpa` fallback). #613 targets `dataflow-up-16-sgl0.5.14` and hasn't been asked to move, so if I retarget #617 alone the diff still carries #613's commits and the stack breaks.
> 
> I checked `colocated`: it already has the same `target_engine` capture-policy layout and the sglang 0.5.14 changes from #652, so #613 should rebase onto it cleanly.
> 
> Maybe #613 retargets to `colocated` first (cc @maocheng23), then I rebase #617 and follow.

Great.
